### PR TITLE
Spirv-fuzz: Enhancing permute function variables and it's testing

### LINKED ISSUE #4257: spirv-fuzz: TranformationPermuteFunctionVariables crashes when the function has no variables
`FuzzerContext::RandomIndex` requires that the vector argument has at least one element.

As a solution, we can check for whether the function has more than 2 variables and apply the transformation only if it does. Additionally, it might be better to limit the number of iterations in the following loop so that we don't end up swapping the same variables (e.g. consider the case when the function has only 2 variables - we'd want to apply at most 1 transformation).

https://github.com/KhronosGroup/SPIRV-Tools/blob/f82f5af5e212373230b8df4dedbd8e1c950e1cc1/source/fuzz/fuzzer_pass_permute_function_variables.cpp#L53-L66

@afd, @Mostafa-ashraf19 ping.

### LINKED ISSUE #4259: spirv-fuzz: Update analyses in TransformationSwapFunctionVariables
When we swap 2 variables, we invalidate DefUseManager. The latter contains a map from a result id to an `opt::Instruction` object. We should update it in the transformation accordingly.

@Mostafa-ashraf19, you can use `opt::DefUseManager::AnalyzeInstDefUse` to do that.
@afd for info.

### PR DESCRIPTION
**Pull request by Mostafa-ashraf19**
Fixes #4257. 
Fixes #4259.

### CODE REVIEW

**Comment by Vasniktel** [comment]
A small note on the PR description. For some reason, GitHub doesn't recognize that the PR will fix 2 issues. I'd suggest changing the description to
```
Fixes #4257
Fixes #4259 
```

**Comment by Mostafa-ashraf19** [comment]
@afd, @Vasniktel, Could you please review this?

File: source/fuzz/transformation_swap_function_variables.cpp [hunk 1]
```diff
@@ -69,6 +69,11 @@ void TransformationSwapFunctionVariables::Apply(
       ir_context->get_def_use_mgr()->GetDef(message_.result_id2());
 
   std::swap(*instruction1, *instruction2);
+
```
Comment (paulthomson) [inline source/fuzz/transformation_swap_function_variables.cpp:72]: After thinking about it, I don't think we should swap the _contents_ of two instructions. That seems difficult in terms of trying to preserve the data structures in ir_context. E.g. the DecorationManager stores pointers to Instruction*, so we really don't want to be changing the pointees. 

Instead, try this:

```
std::unique_ptr<opt::Instruction> temp_instruction =
      MakeUnique<opt::Instruction>();

  temp_instruction->InsertBefore(instruction1);
  instruction1->InsertAfter(instruction2);
  instruction2->InsertAfter(temp_instruction.get());
  temp_instruction->RemoveFromList();
```

With this change, I think you can also remove the lines below that update the def use manager. I don't think we need to update or invalidate any analyses here.

One analysis that you might think you need to invalidate is the dominator analysis. However, the dominates relationship for instructions within a block is computed on-the-fly based on the instruction list, so nothing else needs to be updated or invalidated.

File: test/fuzz/transformation_swap_function_variables_test.cpp [hunk 2]
```diff
@@ -185,24 +186,40 @@ TEST(TransformationSwapFunctionVariables, IsApplicable) {
...
         transformation.IsApplicable(context.get(), transformation_context));
+
     ApplyAndCheckFreshIds(transformation, context.get(),
                           &transformation_context);
+
     ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(
         context.get(), validator_options, kConsoleMessageConsumer));
+
+    ASSERT_EQ(first_instruction, context->get_def_use_mgr()->GetDef(28));
+    ASSERT_EQ(second_instruction, context->get_def_use_mgr()->GetDef(24));
```
Comment (paulthomson) [inline test/fuzz/transformation_swap_function_variables_test.cpp:204]: Once you have made the change above per my suggestion, I guess you will need to swap these back! :-) And the same below.

**Review by paulthomson — COMMENTED** [review]
Looks good! I have left one main suggestion.

File: source/fuzz/fuzzer_pass_permute_function_variables.cpp [hunk 3]
```diff
@@ -49,7 +49,9 @@ void FuzzerPassPermuteFunctionVariables::Apply() {
         variables.push_back(&instruction);
       }
     }
-
+    if (variables.size() <= 1) {
```
[the remark this answers is not in the document]
Comment (Mostafa-ashraf19) [reply source/fuzz/fuzzer_pass_permute_function_variables.cpp:52]: done

[the remark this answers is not in the document]
Comment (Mostafa-ashraf19) [reply source/fuzz/fuzzer_pass_permute_function_variables.cpp:52]: Done.

File: test/fuzz/transformation_swap_function_variables_test.cpp (hunk 2 above)
Comment (Mostafa-ashraf19) [reply test/fuzz/transformation_swap_function_variables_test.cpp:204]: Done.

File: source/fuzz/transformation_swap_function_variables.cpp (hunk 1 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_swap_function_variables.cpp:72]: Done, Thanks for the clarification @paulthomson. Can I understand what is the benefits of a known "dominates relationship for instructions"?

Comment (paulthomson) [reply source/fuzz/transformation_swap_function_variables.cpp:72]: This is explained in the SPIR-V introduction slides, which we will go through again in our meeting...now! :-)

Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_swap_function_variables.cpp:73]: Thanks :-)

**Review by paulthomson — APPROVED** [review]
LGTM!

**Pull request merged**